File: python/tests/tensors/test_LabeledAxis.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


import pytest
from pathlib import Path
import neml2


def test_axis_properties():
    pwd = Path(__file__).parent
    model = neml2.reload_model(pwd / "test_LabeledAxis.i", "model")
    input_axis = model.input_axis()
    output_axis = model.output_axis()

    assert input_axis.has_state()
    assert input_axis.has_old_state()
    assert input_axis.has_forces()
    assert input_axis.has_old_forces()
    assert not input_axis.has_residual()
    assert not input_axis.has_parameters()

    assert not output_axis.has_state()
    assert not output_axis.has_old_state()
    assert not output_axis.has_forces()
    assert not output_axis.has_old_forces()
    assert output_axis.has_residual()
    assert not output_axis.has_parameters()


def test_size():
    pwd = Path(__file__).parent
    model = neml2.reload_model(pwd / "test_LabeledAxis.i", "model")
    input_axis = model.input_axis()
    output_axis = model.output_axis()

    assert input_axis.size() == 8
    assert input_axis.size("forces") == 1
    assert input_axis.size("forces/t") == 1
    assert input_axis.size("old_forces") == 1
    assert input_axis.size("old_forces/t") == 1
    assert input_axis.size("old_state") == 2
    assert input_axis.size("old_state/bar") == 1
    assert input_axis.size("old_state/foo") == 1
    assert input_axis.size("state") == 4
    assert input_axis.size("state/bar") == 1
    assert input_axis.size("state/bar_rate") == 1
    assert input_axis.size("state/foo") == 1
    assert input_axis.size("state/foo_rate") == 1
    assert output_axis.size() == 1
    assert output_axis.size("residual") == 1
    assert output_axis.size("residual/foo_bar") == 1


def test_slice():
    pwd = Path(__file__).parent
    model = neml2.reload_model(pwd / "test_LabeledAxis.i", "model")
    input_axis = model.input_axis()
    output_axis = model.output_axis()

    assert input_axis.slice("forces") == slice(0, 1, 1)
    assert input_axis.slice("forces/t") == slice(0, 1, 1)
    assert input_axis.slice("old_forces") == slice(1, 2, 1)
    assert input_axis.slice("old_forces/t") == slice(1, 2, 1)
    assert input_axis.slice("old_state") == slice(2, 4, 1)
    assert input_axis.slice("old_state/bar") == slice(2, 3, 1)
    assert input_axis.slice("old_state/foo") == slice(3, 4, 1)
    assert input_axis.slice("state") == slice(4, 8, 1)
    assert input_axis.slice("state/bar") == slice(4, 5, 1)
    assert input_axis.slice("state/bar_rate") == slice(5, 6, 1)
    assert input_axis.slice("state/foo") == slice(6, 7, 1)
    assert input_axis.slice("state/foo_rate") == slice(7, 8, 1)
    assert output_axis.slice("residual") == slice(0, 1, 1)
    assert output_axis.slice("residual/foo_bar") == slice(0, 1, 1)


def test_variable_accessors():
    pwd = Path(__file__).parent
    model = neml2.reload_model(pwd / "test_LabeledAxis.i", "model")
    axis = model.input_axis()

    assert axis.nvariable() == 8
    assert axis.has_variable("forces/t")
    assert not axis.has_variable("nonexistent")
    assert axis.variable_id("forces/t") == 0
    assert axis.variable_id("old_forces/t") == 1
    assert axis.variable_id("old_state/bar") == 2
    assert axis.variable_id("old_state/foo") == 3
    assert axis.variable_id("state/bar") == 4
    assert axis.variable_id("state/bar_rate") == 5
    assert axis.variable_id("state/foo") == 6
    assert axis.variable_id("state/foo_rate") == 7
    assert axis.variable_names() == [
        "forces/t",
        "old_forces/t",
        "old_state/bar",
        "old_state/foo",
        "state/bar",
        "state/bar_rate",
        "state/foo",
        "state/foo_rate",
    ]
    assert axis.variable_slices() == [
        slice(0, 1, 1),
        slice(1, 2, 1),
        slice(2, 3, 1),
        slice(3, 4, 1),
        slice(4, 5, 1),
        slice(5, 6, 1),
        slice(6, 7, 1),
        slice(7, 8, 1),
    ]
    assert axis.variable_slice("forces/t") == slice(0, 1, 1)
    assert axis.variable_slice("old_forces/t") == slice(1, 2, 1)
    assert axis.variable_slice("old_state/bar") == slice(2, 3, 1)
    assert axis.variable_slice("old_state/foo") == slice(3, 4, 1)
    assert axis.variable_slice("state/bar") == slice(4, 5, 1)
    assert axis.variable_slice("state/bar_rate") == slice(5, 6, 1)
    assert axis.variable_slice("state/foo") == slice(6, 7, 1)
    assert axis.variable_slice("state/foo_rate") == slice(7, 8, 1)
    assert axis.variable_sizes() == [1, 1, 1, 1, 1, 1, 1, 1]
    assert axis.variable_size("forces/t") == 1
    assert axis.variable_size("old_forces/t") == 1
    assert axis.variable_size("old_state/bar") == 1
    assert axis.variable_size("old_state/foo") == 1
    assert axis.variable_size("state/bar") == 1
    assert axis.variable_size("state/bar_rate") == 1
    assert axis.variable_size("state/foo") == 1
    assert axis.variable_size("state/foo_rate") == 1


def test_subaxis_accessors():
    pwd = Path(__file__).parent
    model = neml2.reload_model(pwd / "test_LabeledAxis.i", "model")
    axis = model.input_axis()

    assert axis.nsubaxis() == 4
    assert axis.has_subaxis("forces")
    assert not axis.has_subaxis("nonexistent")
    assert axis.subaxis_id("forces") == 0
    assert axis.subaxis_id("old_forces") == 1
    assert axis.subaxis_id("old_state") == 2
    assert axis.subaxis_id("state") == 3
    assert axis.subaxis_names() == ["forces", "old_forces", "old_state", "state"]
    assert axis.subaxis_slices() == [
        slice(0, 1, 1),
        slice(1, 2, 1),
        slice(2, 4, 1),
        slice(4, 8, 1),
    ]
    assert axis.subaxis_slice("forces") == slice(0, 1, 1)
    assert axis.subaxis_slice("old_forces") == slice(1, 2, 1)
    assert axis.subaxis_slice("old_state") == slice(2, 4, 1)
    assert axis.subaxis_slice("state") == slice(4, 8, 1)
    assert axis.subaxis_sizes() == [1, 1, 2, 4]
    assert axis.subaxis_size("forces") == 1
    assert axis.subaxis_size("old_forces") == 1
    assert axis.subaxis_size("old_state") == 2
    assert axis.subaxis_size("state") == 4

File: tests/unit/models/test_LabeledAxis.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators_all.hpp>
#include <catch2/matchers/catch_matchers_templated.hpp>

#include "neml2/models/LabeledAxis.h"
#include "neml2/tensors/tensors.h"

using namespace neml2;

struct EqualsSlice : Catch::Matchers::MatcherGenericBase
{
  EqualsSlice(const indexing::Slice & slice)
    : slice{slice}
  {
  }

  bool match(const indexing::Slice & other) const
  {
    return slice.start() == other.start() && slice.stop() == other.stop() &&
           slice.step() == other.step();
  }

  std::string describe() const override { return "Equals: " + utils::stringify(slice); }

private:
  const indexing::Slice & slice;
};

TEST_CASE("LabeledAxis", "[models]")
{
  SECTION("class LabeledAxis")
  {
    LabeledAxis a;
    a.add_variable<Scalar>("scalar");
    a.add_variable<SR2>("r2t");
    a.add_subaxis("sub1");
    a.subaxis("sub1").add_variable<Scalar>("scalar");
    a.subaxis("sub1").add_variable<SR2>("r2t");
    a.subaxis("sub1").add_subaxis("sub2");
    a.subaxis("sub1").subaxis("sub2").add_variable<Scalar>("scalar");
    a.subaxis("sub1").subaxis("sub2").add_variable<SR2>("r2t");
    a.subaxis("sub1").add_subaxis("sub3");
    a.subaxis("sub1").subaxis("sub3").add_variable<Scalar>("scalar");
    a.subaxis("sub1").subaxis("sub3").add_variable("foo", 5);
    const auto & sub1 = a.subaxis("sub1");
    const auto & sub2 = sub1.subaxis("sub2");
    const auto & sub3 = sub1.subaxis("sub3");

    auto setup = GENERATE(true, false);

    if (setup)
      a.setup_layout();

    SECTION("is_setup") { REQUIRE(a.is_setup() == setup); }

    SECTION("size")
    {
      REQUIRE(a.size() == 27);
      REQUIRE(a.size("r2t") == 6);
      REQUIRE(a.size("scalar") == 1);
      REQUIRE(a.size({"sub1", "r2t"}) == 6);
      REQUIRE(a.size({"sub1", "scalar"}) == 1);
      REQUIRE(a.size({"sub1", "sub2", "r2t"}) == 6);
      REQUIRE(a.size({"sub1", "sub2", "scalar"}) == 1);
      REQUIRE(a.size({"sub1", "sub3", "foo"}) == 5);
      REQUIRE(a.size({"sub1", "sub3", "scalar"}) == 1);
      REQUIRE(a.size("sub1") == 20);
      REQUIRE(a.size({"sub1", "sub2"}) == 7);
      REQUIRE(a.size({"sub1", "sub3"}) == 6);

      REQUIRE(sub1.size() == 20);
      REQUIRE(sub1.size("r2t") == 6);
      REQUIRE(sub1.size("scalar") == 1);
      REQUIRE(sub1.size({"sub2", "r2t"}) == 6);
      REQUIRE(sub1.size({"sub2", "scalar"}) == 1);
      REQUIRE(sub1.size({"sub3", "foo"}) == 5);
      REQUIRE(sub1.size({"sub3", "scalar"}) == 1);
      REQUIRE(sub1.size("sub2") == 7);
      REQUIRE(sub1.size("sub3") == 6);

      REQUIRE(sub2.size() == 7);
      REQUIRE(sub2.size("r2t") == 6);
      REQUIRE(sub2.size("scalar") == 1);

      REQUIRE(sub3.size() == 6);
      REQUIRE(sub3.size("foo") == 5);
      REQUIRE(sub3.size("scalar") == 1);
    }

    SECTION("slice")
    {
      if (setup)
      {
        REQUIRE_THAT(a.slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(a.slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(a.slice({"sub1", "r2t"}), EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(a.slice({"sub1", "scalar"}), EqualsSlice(indexing::Slice(13, 14)));
        REQUIRE_THAT(a.slice({"sub1", "sub2", "r2t"}), EqualsSlice(indexing::Slice(14, 20)));
        REQUIRE_THAT(a.slice({"sub1", "sub2", "scalar"}), EqualsSlice(indexing::Slice(20, 21)));
        REQUIRE_THAT(a.slice({"sub1", "sub3", "foo"}), EqualsSlice(indexing::Slice(21, 26)));
        REQUIRE_THAT(a.slice({"sub1", "sub3", "scalar"}), EqualsSlice(indexing::Slice(26, 27)));
        REQUIRE_THAT(a.slice("sub1"), EqualsSlice(indexing::Slice(7, 27)));
        REQUIRE_THAT(a.slice({"sub1", "sub2"}), EqualsSlice(indexing::Slice(14, 21)));
        REQUIRE_THAT(a.slice({"sub1", "sub3"}), EqualsSlice(indexing::Slice(21, 27)));

        REQUIRE_THAT(sub1.slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub1.slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(sub1.slice({"sub2", "r2t"}), EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(sub1.slice({"sub2", "scalar"}), EqualsSlice(indexing::Slice(13, 14)));
        REQUIRE_THAT(sub1.slice({"sub3", "foo"}), EqualsSlice(indexing::Slice(14, 19)));
        REQUIRE_THAT(sub1.slice({"sub3", "scalar"}), EqualsSlice(indexing::Slice(19, 20)));
        REQUIRE_THAT(sub1.slice("sub2"), EqualsSlice(indexing::Slice(7, 14)));
        REQUIRE_THAT(sub1.slice("sub3"), EqualsSlice(indexing::Slice(14, 20)));

        REQUIRE_THAT(sub2.slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub2.slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));

        REQUIRE_THAT(sub3.slice("foo"), EqualsSlice(indexing::Slice(0, 5)));
        REQUIRE_THAT(sub3.slice("scalar"), EqualsSlice(indexing::Slice(5, 6)));
      }
    }

    SECTION("nvariable")
    {
      REQUIRE(a.nvariable() == 8);
      REQUIRE(sub1.nvariable() == 6);
      REQUIRE(sub2.nvariable() == 2);
      REQUIRE(sub3.nvariable() == 2);
    }

    SECTION("has_variable")
    {
      REQUIRE(a.has_variable("r2t"));
      REQUIRE(a.has_variable("scalar"));
      REQUIRE(!a.has_variable("foo"));
      REQUIRE(sub1.has_variable("r2t"));
      REQUIRE(!sub1.has_variable("foo"));
    }

    SECTION("variable_id")
    {
      if (setup)
      {
        REQUIRE(a.variable_id("r2t") == 0);
        REQUIRE(a.variable_id("scalar") == 1);
        REQUIRE(a.variable_id({"sub1", "r2t"}) == 2);
        REQUIRE(a.variable_id({"sub1", "scalar"}) == 3);
        REQUIRE(a.variable_id({"sub1", "sub2", "r2t"}) == 4);
        REQUIRE(a.variable_id({"sub1", "sub2", "scalar"}) == 5);
        REQUIRE(a.variable_id({"sub1", "sub3", "foo"}) == 6);
        REQUIRE(a.variable_id({"sub1", "sub3", "scalar"}) == 7);

        REQUIRE(sub1.variable_id("r2t") == 0);
        REQUIRE(sub1.variable_id("scalar") == 1);
        REQUIRE(sub1.variable_id({"sub2", "r2t"}) == 2);
        REQUIRE(sub1.variable_id({"sub2", "scalar"}) == 3);
        REQUIRE(sub1.variable_id({"sub3", "foo"}) == 4);
        REQUIRE(sub1.variable_id({"sub3", "scalar"}) == 5);

        REQUIRE(sub2.variable_id("r2t") == 0);
        REQUIRE(sub2.variable_id("scalar") == 1);

        REQUIRE(sub3.variable_id("foo") == 0);
        REQUIRE(sub3.variable_id("scalar") == 1);
      }
    }

    SECTION("variable_names")
    {
      if (setup)
      {
        REQUIRE(a.variable_names() == std::vector<LabeledAxisAccessor>{"r2t",
                                                                       "scalar",
                                                                       {"sub1", "r2t"},
                                                                       {"sub1", "scalar"},
                                                                       {"sub1", "sub2", "r2t"},
                                                                       {"sub1", "sub2", "scalar"},
                                                                       {"sub1", "sub3", "foo"},
                                                                       {"sub1", "sub3", "scalar"}});
        REQUIRE(sub1.variable_names() == std::vector<LabeledAxisAccessor>{"r2t",
                                                                          "scalar",
                                                                          {"sub2", "r2t"},
                                                                          {"sub2", "scalar"},
                                                                          {"sub3", "foo"},
                                                                          {"sub3", "scalar"}});
        REQUIRE(sub2.variable_names() == std::vector<LabeledAxisAccessor>{"r2t", "scalar"});
        REQUIRE(sub3.variable_names() == std::vector<LabeledAxisAccessor>{"foo", "scalar"});
      }
    }

    SECTION("variable_slices")
    {
      if (setup)
      {
        REQUIRE_THAT(a.variable_slices()[0], EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(a.variable_slices()[1], EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(a.variable_slices()[2], EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(a.variable_slices()[3], EqualsSlice(indexing::Slice(13, 14)));
        REQUIRE_THAT(a.variable_slices()[4], EqualsSlice(indexing::Slice(14, 20)));
        REQUIRE_THAT(a.variable_slices()[5], EqualsSlice(indexing::Slice(20, 21)));
        REQUIRE_THAT(a.variable_slices()[6], EqualsSlice(indexing::Slice(21, 26)));
        REQUIRE_THAT(a.variable_slices()[7], EqualsSlice(indexing::Slice(26, 27)));

        REQUIRE_THAT(sub1.variable_slices()[0], EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub1.variable_slices()[1], EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(sub1.variable_slices()[2], EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(sub1.variable_slices()[3], EqualsSlice(indexing::Slice(13, 14)));

        REQUIRE_THAT(sub2.variable_slices()[0], EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub2.variable_slices()[1], EqualsSlice(indexing::Slice(6, 7)));

        REQUIRE_THAT(sub3.variable_slices()[0], EqualsSlice(indexing::Slice(0, 5)));
        REQUIRE_THAT(sub3.variable_slices()[1], EqualsSlice(indexing::Slice(5, 6)));
      }
    }

    SECTION("variable_slice")
    {
      if (setup)
      {
        REQUIRE_THAT(a.variable_slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(a.variable_slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(a.variable_slice({"sub1", "r2t"}), EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(a.variable_slice({"sub1", "scalar"}), EqualsSlice(indexing::Slice(13, 14)));
        REQUIRE_THAT(a.variable_slice({"sub1", "sub2", "r2t"}),
                     EqualsSlice(indexing::Slice(14, 20)));
        REQUIRE_THAT(a.variable_slice({"sub1", "sub2", "scalar"}),
                     EqualsSlice(indexing::Slice(20, 21)));
        REQUIRE_THAT(a.variable_slice({"sub1", "sub3", "foo"}),
                     EqualsSlice(indexing::Slice(21, 26)));
        REQUIRE_THAT(a.variable_slice({"sub1", "sub3", "scalar"}),
                     EqualsSlice(indexing::Slice(26, 27)));

        REQUIRE_THAT(sub1.variable_slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub1.variable_slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));
        REQUIRE_THAT(sub1.variable_slice({"sub2", "r2t"}), EqualsSlice(indexing::Slice(7, 13)));
        REQUIRE_THAT(sub1.variable_slice({"sub2", "scalar"}), EqualsSlice(indexing::Slice(13, 14)));
        REQUIRE_THAT(sub1.variable_slice({"sub3", "foo"}), EqualsSlice(indexing::Slice(14, 19)));
        REQUIRE_THAT(sub1.variable_slice({"sub3", "scalar"}), EqualsSlice(indexing::Slice(19, 20)));

        REQUIRE_THAT(sub2.variable_slice("r2t"), EqualsSlice(indexing::Slice(0, 6)));
        REQUIRE_THAT(sub2.variable_slice("scalar"), EqualsSlice(indexing::Slice(6, 7)));

        REQUIRE_THAT(sub3.variable_slice("foo"), EqualsSlice(indexing::Slice(0, 5)));
        REQUIRE_THAT(sub3.variable_slice("scalar"), EqualsSlice(indexing::Slice(5, 6)));
      }
    }

    SECTION("variable_sizes")
    {
      if (setup)
      {
        REQUIRE(a.variable_sizes() == std::vector<Size>{6, 1, 6, 1, 6, 1, 5, 1});
        REQUIRE(sub1.variable_sizes() == std::vector<Size>{6, 1, 6, 1, 5, 1});
        REQUIRE(sub2.variable_sizes() == std::vector<Size>{6, 1});
        REQUIRE(sub3.variable_sizes() == std::vector<Size>{5, 1});
      }
    }

    SECTION("variable_size")
    {
      REQUIRE(a.variable_size("r2t") == 6);
      REQUIRE(a.variable_size("scalar") == 1);
      REQUIRE(a.variable_size({"sub1", "r2t"}) == 6);
      REQUIRE(a.variable_size({"sub1", "scalar"}) == 1);
      REQUIRE(a.variable_size({"sub1", "sub2", "r2t"}) == 6);
      REQUIRE(a.variable_size({"sub1", "sub2", "scalar"}) == 1);
      REQUIRE(a.variable_size({"sub1", "sub3", "foo"}) == 5);
      REQUIRE(a.variable_size({"sub1", "sub3", "scalar"}) == 1);

      REQUIRE(sub1.variable_size("r2t") == 6);
      REQUIRE(sub1.variable_size("scalar") == 1);
      REQUIRE(sub1.variable_size({"sub2", "r2t"}) == 6);
      REQUIRE(sub1.variable_size({"sub2", "scalar"}) == 1);
      REQUIRE(sub1.variable_size({"sub3", "foo"}) == 5);
      REQUIRE(sub1.variable_size({"sub3", "scalar"}) == 1);

      REQUIRE(sub2.variable_size("r2t") == 6);
      REQUIRE(sub2.variable_size("scalar") == 1);

      REQUIRE(sub3.variable_size("foo") == 5);
      REQUIRE(sub3.variable_size("scalar") == 1);
    }

    SECTION("nsubaxis")
    {
      REQUIRE(a.nsubaxis() == 1);
      REQUIRE(sub1.nsubaxis() == 2);
      REQUIRE(sub2.nsubaxis() == 0);
      REQUIRE(sub3.nsubaxis() == 0);
    }

    SECTION("has_subaxis")
    {
      REQUIRE(a.has_subaxis("sub1"));
      REQUIRE(a.has_subaxis({"sub1", "sub2"}));
      REQUIRE(a.has_subaxis({"sub1", "sub3"}));
      REQUIRE(!a.has_subaxis("sub2"));
      REQUIRE(sub1.has_subaxis("sub2"));
      REQUIRE(sub1.has_subaxis("sub3"));
      REQUIRE(!sub1.has_subaxis("sub4"));
    }

    SECTION("subaxis_id")
    {
      if (setup)
      {
        REQUIRE(a.subaxis_id("sub1") == 0);
        REQUIRE(sub1.subaxis_id("sub2") == 0);
        REQUIRE(sub1.subaxis_id("sub3") == 1);
      }
    }

    SECTION("subaxes")
    {
      if (setup)
      {
        REQUIRE(a.subaxes() == std::vector<const LabeledAxis *>{&sub1});
        REQUIRE(sub1.subaxes() == std::vector<const LabeledAxis *>{&sub2, &sub3});
      }
    }

    SECTION("subaxis")
    {
      REQUIRE(&a.subaxis("sub1") == &sub1);
      REQUIRE(&a.subaxis({"sub1", "sub2"}) == &sub2);
      REQUIRE(&a.subaxis({"sub1", "sub3"}) == &sub3);
      REQUIRE(&sub1.subaxis("sub2") == &sub2);
      REQUIRE(&sub1.subaxis("sub3") == &sub3);
    }

    SECTION("subaxis_names")
    {
      if (setup)
      {
        REQUIRE(a.subaxis_names() == std::vector<std::string>{"sub1"});
        REQUIRE(sub1.subaxis_names() == std::vector<std::string>{"sub2", "sub3"});
      }
    }

    SECTION("subaxis_slices")
    {
      if (setup)
      {
        REQUIRE_THAT(a.subaxis_slices()[0], EqualsSlice(indexing::Slice(7, 27)));
        REQUIRE_THAT(sub1.subaxis_slices()[0], EqualsSlice(indexing::Slice(7, 14)));
        REQUIRE_THAT(sub1.subaxis_slices()[1], EqualsSlice(indexing::Slice(14, 20)));
      }
    }

    SECTION("subaxis_slice")
    {
      if (setup)
      {
        REQUIRE_THAT(a.subaxis_slice("sub1"), EqualsSlice(indexing::Slice(7, 27)));
        REQUIRE_THAT(a.subaxis_slice({"sub1", "sub2"}), EqualsSlice(indexing::Slice(14, 21)));
        REQUIRE_THAT(a.subaxis_slice({"sub1", "sub3"}), EqualsSlice(indexing::Slice(21, 27)));
        REQUIRE_THAT(sub1.subaxis_slice("sub2"), EqualsSlice(indexing::Slice(7, 14)));
        REQUIRE_THAT(sub1.subaxis_slice("sub3"), EqualsSlice(indexing::Slice(14, 20)));
      }
    }

    SECTION("subaxis_sizes")
    {
      if (setup)
      {
        REQUIRE(a.subaxis_sizes() == std::vector<Size>{20});
        REQUIRE(sub1.subaxis_sizes() == std::vector<Size>{7, 6});
      }
    }

    SECTION("subaxis_size")
    {
      REQUIRE(a.subaxis_size("sub1") == 20);
      REQUIRE(a.subaxis_size({"sub1", "sub2"}) == 7);
      REQUIRE(a.subaxis_size({"sub1", "sub3"}) == 6);
      REQUIRE(sub1.subaxis_size("sub2") == 7);
      REQUIRE(sub1.subaxis_size("sub3") == 6);
    }
  }

  SECTION("operator==")
  {
    // test1 and test2 are equal:
    // They have the same set of items with same names and storage sizes.
    LabeledAxis test1;
    test1.add_variable<Scalar>("scalar1");
    test1.add_variable<SR2>("r2t1");
    test1.add_variable<Scalar>("scalar2");
    test1.add_variable<SR2>("r2t2");
    test1.add_variable<Scalar>("scalar3");
    test1.setup_layout();

    LabeledAxis test2;
    test2.add_variable<SR2>("r2t1");
    test2.add_variable<Scalar>("scalar3");
    test2.add_variable<SR2>("r2t2");
    test2.add_variable<Scalar>("scalar2");
    test2.add_variable<Scalar>("scalar1");
    test2.setup_layout();

    REQUIRE(test1 == test2);
  }

  SECTION("operator!=")
  {
    // test1 and test2 are equal:
    // They have the same set of items with same names and storage sizes.
    LabeledAxis test1;
    test1.add_variable<Scalar>("scalar1");
    test1.add_variable<SR2>("r2t1");
    test1.add_variable<Scalar>("scalar2");
    test1.add_variable<SR2>("r2t2");
    test1.add_variable<Scalar>("scalar3");
    test1.setup_layout();

    LabeledAxis test2;
    test2.add_variable<SR2>("r2t1");
    test2.add_variable<Scalar>("scalar3");
    test2.add_variable<SR2>("r2t2");
    test2.add_variable<Scalar>("scalar2");
    test2.add_variable<Scalar>("scalar1");
    test2.setup_layout();

    // test3 is NOT equal to test1 nor test2
    LabeledAxis test3;
    test3.add_variable<SR2>("r2t1");
    test3.add_variable<Scalar>("scalar3");
    test3.add_variable<SR2>("r2t2");
    test3.setup_layout();

    REQUIRE(test1 != test3);
    REQUIRE(test2 != test3);
  }
}
